File: monitor_training.R
#
# Plot training progress
#
# (c) 2021 Jean-Olivier Irisson, GNU General Public License v3

library("tidyverse")
library("patchwork")

# read training log
input <- "io/checkpoints/training_log.tsv"
d <- read_tsv(input, col_types = cols())
dm <- d %>%
  # reformat to get losses and accuracies in the same column
  pivot_longer(cols=c(starts_with("train"), starts_with("val"))) %>%
  separate(name, into=c("dataset", "metric"), sep="_") %>%
  pivot_wider(names_from=metric, values_from=value)
de <- filter(d, batch == max(batch))

# define plots
base <- ggplot() + geom_vline(aes(xintercept=step), data=de, colour="grey80") + theme(legend.position="top")
p_loss <- base + geom_path(aes(x=step, y=loss, colour=dataset), data=dm)
p_accu <- base + geom_path(aes(x=step, y=accuracy, colour=dataset), data=dm)
p_lr <- base + geom_path(aes(x=step, y=learning_rate), data=d)

# plot in a nice grid
(p_loss + p_accu) / p_lr + plot_layout(heights=c(2.5,1))

File: monitor_training.py
#!/usr/bin/env python
#
# Read the training log and plot the current stats
#
# (c) 2023 Jean-Olivier Irisson, GNU General Public License v3

import pandas as pd
from matplotlib import pyplot as plt

df = pd.read_csv('io/checkpoints/training_log.tsv', sep='\t')
df = df.drop(['batch', 'learning_rate'], axis='columns')

df.plot(x='step', subplots=True)
plt.show()

df.plot(x='epoch', subplots=True)
plt.show()
